### DISCUSSION TITLE
Dynamic CSV file as BC

### DISCUSSION THREAD

**Discussion by Ghazipo** [idaholab/moose#25390] [Q&A General]
Dear MOOSE community

I'm encountering an issue with setting a dynamic CSV file as a Boundary Condition (BC) in my application. In my setup, I'm using a FunctionDirichletBC that calls a function of type PiecewiseConstant, and this function is configured to read data from a CSV file. The challenge is that this CSV file is expected to change over time during the simulation.

I've implemented an external code that appends a new row to the CSV file after each time step of the MOOSE simulation. However, my MOOSE-based application appears to only consider the initial version of the CSV file before the start of the simulation and doesn't recognize its dynamic changes during the simulation.

I attempted to address this issue by changing the execute_on option in the Function block to "TIMESTEP_BEGIN," with the expectation that MOOSE would recognize the updated CSV file at the beginning of each timestep. Surprisingly, this approach didn't yield the desired results. Strangely, the behavior remains the same whether the execute_on option is set to "none" or "TIMESTEP_BEGIN." 

To be clear and concise, I would greatly appreciate any insights into how MOOSE handles the variations in a CSV file during a simulation. What methods or configurations can be used to ensure that MOOSE recognizes and applies the changes in the CSV file during the simulation?

Thank you for your assistance.

**Comment by GiudGiud** [comment]
Hello

The file is read at construction in that function
https://github.com/idaholab/moose/blob/next/framework/src/functions/PiecewiseTabularBase.C

If you want it to be read on every step, you ll need to copy this code to a timestepSetup routine to get the file to be read again
```
  // determine data source and check parameter consistency
  if (isParamValid("data_file") && !isParamValid("x") && !isParamValid("y") &&
      !isParamValid("xy_data"))
    buildFromFile();
  else if (!isParamValid("data_file") && isParamValid("x") && isParamValid("y") &&
           !isParamValid("xy_data"))
    buildFromXandY();
  else if (!isParamValid("data_file") && !isParamValid("x") && !isParamValid("y") &&
           isParamValid("xy_data"))
    buildFromXY();
  else
    mooseError("In ",
               _name,
               ": Either 'data_file', 'x' and 'y', or 'xy_data' must be specified exclusively.");
```
You really just need to move the buildFromFile() routine

Guillaume

**Comment by GiudGiud** [comment]
Hello

The file is read at construction in that function
https://github.com/idaholab/moose/blob/next/framework/src/functions/PiecewiseTabularBase.C

If you want it to be read on every step, you ll need to copy this code to a timestepSetup routine to get the file to be read again
```
  // determine data source and check parameter consistency
  if (isParamValid("data_file") && !isParamValid("x") && !isParamValid("y") &&
      !isParamValid("xy_data"))
    buildFromFile();
  else if (!isParamValid("data_file") && isParamValid("x") && isParamValid("y") &&
           !isParamValid("xy_data"))
    buildFromXandY();
  else if (!isParamValid("data_file") && !isParamValid("x") && !isParamValid("y") &&
           isParamValid("xy_data"))
    buildFromXY();
  else
    mooseError("In ",
               _name,
               ": Either 'data_file', 'x' and 'y', or 'xy_data' must be specified exclusively.");
```
You really just need to move the buildFromFile() routine

Guillaume
